lcls-twincat-motion/Library/POUs/Motion/FB_GantryAutoCoupling.TcPOU

```diff
@@ -0,0 +1,89 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<TcPlcObject Version="1.1.0.1" ProductVersion="3.1.4020.14">
+  <POU Name="FB_GantryAutoCoupling" Id="{532dfcaf-6187-4140-aaad-e20a9bbe6b43}" SpecialFunc="None">
+    <Declaration><![CDATA[FUNCTION_BLOCK FB_GantryAutoCoupling
+VAR_INPUT
+	fGantryTol : LREAL;
+END_VAR
+VAR_OUTPUT
+	bGantryAlreadyCoupled : BOOL;
+END_VAR
+VAR_IN_OUT
+	Master : DUT_MotionStage;
+	MasterEnc : ST_RenishawAbsEnc;
+	Slave : DUT_MotionStage;
+	SlaveEnc : ST_RenishawAbsEnc;
+	bExecuteCouple : BOOL;
+	bExecuteDecouple : BOOL;
+END_VAR
+VAR
+	gantry_diff_limit : FB_GantryDiffVirtualLimitSwitch;
+	couple : MC_GEARIN;
+	decouple : MC_GEAROUT;
+
+
+	bInitComplete : BOOL;
+END_VAR
+]]></Declaration>
+    <Implementation>
+      <ST><![CDATA[// Designate Master and SLave Axes
+Master.bGantryAxis := TRUE;
+Slave.bGantryAxis := TRUE;
+
+Master.fGantryTol := fGantryTol;
+Slave.fGantryTol := Master.fGantryTol;
+
+// Activate Gantry Virtual Limit Switch
+gantry_diff_limit(Penc:=MasterEnc, SEnc:=SlaveEnc, GantDiffTol:=LREAL_TO_LINT(Master.fGantryTol),
+				  PLimFwd=>Master.bLimGntFwd, PLimBwd=>Master.bLimGntBwd,
+				  SLimFwd=>Slave.bLimGntFwd, SLimBwd=>Slave.bLimGntBwd);
+
+// Coupling Status Bit
+bGantryAlreadyCoupled := Master.Axis.NcToPlc.CoupleState=1 AND Slave.Axis.NcToPlc.CoupleState=3;
+
+// Make master and slave axes respect each other's limits
+IF bGantryAlreadyCoupled THEN
+	// Fwd limits
+	Master.bLimGntFwd := Master.bLimGntFwd AND Slave.bLimGntFwd AND Slave.bLimFwd;
+	Slave.bLimGntFwd := Master.bLimGntFwd AND Slave.bLimGntFwd AND Master.bLimFwd;
+	// Bwd limits
+	Master.bLimGntBwd := Master.bLimGntBwd AND Slave.bLimGntBwd AND Slave.bLimBwd;
+	Slave.bLimGntBwd := Master.bLimGntBwd AND Slave.bLimGntBwd AND Master.bLimBwd;
+END_IF
+
+// Coupling states
+// Auto-coupling at init and auto-reset of coupling boolean
+bExecuteCouple S= NOT bInitComplete;
```

Comment:
I started reviewing but got stuck on these `S=` and `R=` operators. What do these operators do?

Reply:
Oh yes Alex showed those to me. The command `S=` reads as "Set by" so for example `A S= B` means when B goes to True A also goes to True but not vice-versa. `R=` reads as "Reset by" and does the opposite so `A R= B` means when B goes to true A goes back to false.

So the logic here is:
1) `bExecuteCouple S= NOT bInitComplete;`
2) `bExecuteCouple R= couple.Busy OR bGantryAlreadyCoupled;`
3) `couple(Master:=Master.Axis, Slave:=Slave.Axis, Execute:=bExecuteCouple);`
4) `bInitComplete S= bGantryAlreadyCoupled;`
5) `bExecuteDecouple R= decouple.Busy OR NOT bGantryAlreadyCoupled;`
6) `decouple(Slave:=Slave.Axis, Execute:=bExecuteDecouple);`

Since all bools start False, on startup (1) makes `bExecuteCouple = True` which makes the `couple` function block execute. Then once coupling is confirmed by `bGantryAlreadyCoupled`, `bInitComplete` will go to `True` which causes the `couple` FB to not be repeatedly executed. Then when the user wants to decouple they write `bExecuteDecouple` to `True` which makes line (6) execute, and then this bool resets back to `False` by line (5) after the axes are no longer coupled

Reply:
Ok, I understand now, thanks. These are missing from Beckhoff's documentation on structured text operators, but they've left two rows for them. See https://infosys.beckhoff.com/english.php?content=../content/1033/tcplccontrol/html/tcplcctrl_iec_operators_overview.htm&id=2174038239638951926, two blank boxes where there should be `S=` and `R=`.

Reply:
Those S/R operators are interesting. I wonder if it means "set" and "reset" as in an "SR flip-flop" - or more verbosely in this case "positive edge-triggered set and reset", I suppose.